apps/passport-server/src/util/telegramMenu.ts

```diff
@@ -0,0 +1,145 @@
+  return ctx.from === undefined || ctx.chat === undefined
+    ? undefined
+    : `${ctx.from.id}/${ctx.chat.id}`;
+};
+
+const checkDeleteMessage = (ctx: BotContext): void => {
+  if (ctx.chat?.id && ctx.session?.lastMessageId) {
+    ctx.api.deleteMessage(ctx.chat.id, ctx.session.lastMessageId);
+    ctx.session.lastMessageId = 0;
+  }
+};
+
+const editOrSendMessage = async (
+  ctx: BotContext,
+  replyText: string
+): Promise<void> => {
+  if (ctx.chat?.id && ctx.session.lastMessageId) {
+    await ctx.api.editMessageText(
+      ctx.chat?.id,
+      ctx.session.lastMessageId,
+      "...",
+      {
+        parse_mode: "HTML"
+      }
+    );
+    await ctx.api.editMessageText(
+      ctx.chat?.id,
+      ctx.session.lastMessageId,
+      replyText,
+      {
+        parse_mode: "HTML"
+      }
+    );
+  } else {
+    const msg = await ctx.reply(replyText, { parse_mode: "HTML" });
+    ctx.session.lastMessageId = msg.message_id;
+  }
+};
+
+export const dynamicEvents = async (
+  ctx: BotContext,
+  range: MenuRange<BotContext>
+): Promise<void> => {
+  const db = ctx.session.dbPool;
+  if (!db) {
+    range.text(`Database not connected. Try again...`);
+    return;
+  }
+  // If an event is selected, display it and its menu options
+  if (ctx.session.selectedEvent) {
+    const event = ctx.session.selectedEvent;
+
+    range.text(`${event.isLinked ? "✅" : ""} ${event.eventName}`).row();
+    range
+      .text(`Yes, ${event.isLinked ? "Remove" : "Add"}`, async (ctx) => {
+        let replyText = "";
+        if (!(await isAdmin(ctx))) return;
+
+        if (!ctx.chat?.id) {
+          await editOrSendMessage(ctx, `Chat Id not found`);
+        } else {
+          if (!event.isLinked) {
+            replyText = `<i>Added ${event.eventName} to chat</i>`;
+            await insertTelegramEvent(db, event.configEventID, ctx.chat.id);
+            await editOrSendMessage(ctx, replyText);
+          } else {
+            replyText = `<i>Removed ${event.eventName} to chat</i>`;
+            await deleteTelegramEvent(db, event.configEventID);
+          }
+        }
+        ctx.session.selectedEvent = undefined;
+        await ctx.menu.update({ immediate: true });
+        await editOrSendMessage(ctx, replyText);
+      })
+      .row();
+
+    range.text(`Go back`, async (ctx) => {
+      if (!(await isAdmin(ctx))) return;
+      checkDeleteMessage(ctx);
+      ctx.session.selectedEvent = undefined;
+      await ctx.menu.update({ immediate: true });
+    });
+  }
+  // Otherwise, display all events to manage.
+  else {
+    const events = await fetchLinkedPretixAndTelegramEvents(db);
+    const eventsWithGateStatus = events.map((e) => {
+      return { ...e, isLinked: e.telegramChatID === ctx.chat?.id.toString() };
+    });
+    for (const event of eventsWithGateStatus) {
+      range
+        .text(
+          `${event.isLinked ? "✅" : ""} ${event.eventName}`,
+          async (ctx) => {
+            if (!(await isAdmin(ctx))) return;
```

Comment:
should we send a message if they're not an admin?

Reply:
I don't think we should send a message below the menu, but perhaps we can send a callbackQuery at the top

Reply:
wasn't able to make this work, will add to a future todo in the interest of testing the auth flow